crates/uv-client/src/base_client.rs

```diff
@@ -1139,8 +1139,66 @@ fn retryable_on_request_failure(err: &(dyn Error + 'static)) -> Option<Retryable
     None
 }
 
-pub fn is_transient_network_error(err: &(dyn Error + 'static)) -> bool {
-    retryable_on_request_failure(err) == Some(Retryable::Transient)
+/// Decide whether a requests should be retried, tracking previous attempts and backoff.
+pub struct RetryState {
+    retry_policy: ExponentialBackoff,
+    start_time: SystemTime,
+    total_retries: u32,
+    url: DisplaySafeUrl,
+}
+
+impl RetryState {
+    pub fn new(retry_policy: ExponentialBackoff, url: impl Into<DisplaySafeUrl>) -> Self {
+        Self {
+            retry_policy,
+            start_time: SystemTime::now(),
+            total_retries: 0,
+            url: url.into(),
+        }
+    }
+
+    /// The number of retries across all requests.
+    ///
+    /// After a failed retryable request, this equals the maximum number of retries.
+    pub fn total_retries(&self) -> u32 {
+        self.total_retries
+    }
+
+    /// Whether request should be retried. Waits with backoff if required.
+    ///
+    /// Takes the number of retries from nested layers associated with the specific `err` type as
+    /// `error_retries`.
+    pub async fn should_retry(&mut self, err: &(dyn Error + 'static), error_retries: u32) -> bool {
+        // Check if the middleware already performed retries
+        self.total_retries += error_retries;
+        match retryable_on_request_failure(err) {
+            Some(Retryable::Transient) => {
+                // If middleware already retried, consider that in our retry budget
```

Comment:
I still think this comment is confusing as the point at which we are "considering" the upstream retries is when we add them. I would just drop it.

```suggestion
```